tools/database/src/analyse_gas_usage.rs

```diff
@@ -0,0 +1,348 @@
+            }
+        };
+
+        // Analyse
+        analyse_gas_usage(blocks_iter, &chain_store, &epoch_manager);
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct GasUsageInShard {
+    pub used_gas_per_account: BTreeMap<AccountId, BigGas>,
+    pub used_gas_total: BigGas,
+}
+
+/// A shard can be split into two halves.
+/// This struct represents the result of splitting a shard at `split_account`.
+#[derive(Debug, Clone)]
+struct ShardSplit {
+    /// Account on which the shard would be split
+    pub split_account: AccountId,
+    /// Gas used by accounts < split_account
+    pub gas_left: BigGas,
+    /// Gas used by accounts >= split_account
+    pub gas_right: BigGas,
+}
+
+impl GasUsageInShard {
+    pub fn new() -> GasUsageInShard {
+        GasUsageInShard { used_gas_per_account: BTreeMap::new(), used_gas_total: 0 }
+    }
+
+    pub fn add_used_gas(&mut self, account: AccountId, used_gas: BigGas) {
+        let old_used_gas: &BigGas = self.used_gas_per_account.get(&account).unwrap_or(&0);
+        let new_used_gas: BigGas = old_used_gas.checked_add(used_gas).unwrap();
+        self.used_gas_per_account.insert(account, new_used_gas);
+
+        self.used_gas_total = self.used_gas_total.checked_add(used_gas).unwrap();
+    }
+
+    pub fn merge(&mut self, other: &GasUsageInShard) {
+        for (account_id, used_gas) in &other.used_gas_per_account {
+            self.add_used_gas(account_id.clone(), *used_gas);
+        }
+        self.used_gas_total = self.used_gas_total.checked_add(other.used_gas_total).unwrap();
+    }
+
+    /// Calculate the optimal point at which this shard could be split into two halves with similar gas usage
+    pub fn calculate_split(&self) -> Option<ShardSplit> {
+        let mut split_account = match self.used_gas_per_account.keys().next() {
+            Some(account_id) => account_id,
+            None => return None,
+        };
+
+        if self.used_gas_per_account.len() < 2 {
+            return None;
+        }
+
+        let mut gas_left: BigGas = 0;
+        let mut gas_right: BigGas = self.used_gas_total;
+
+        for (account, used_gas) in self.used_gas_per_account.iter() {
+            if gas_left >= gas_right {
+                break;
+            }
+
+            split_account = &account;
+            gas_left = gas_left.checked_add(*used_gas).unwrap();
+            gas_right = gas_right.checked_sub(*used_gas).unwrap();
+        }
+
+        Some(ShardSplit { split_account: split_account.clone(), gas_left, gas_right })
+    }
+}
+
+#[derive(Clone, Debug)]
+struct GasUsageStats {
+    pub shards: BTreeMap<ShardUId, GasUsageInShard>,
+}
+
+impl GasUsageStats {
+    pub fn new() -> GasUsageStats {
+        GasUsageStats { shards: BTreeMap::new() }
+    }
+
+    pub fn add_gas_usage_in_shard(&mut self, shard_uid: ShardUId, shard_usage: GasUsageInShard) {
+        match self.shards.get_mut(&shard_uid) {
```

Comment:
Using the `enrty()` api it would look like this:
```rust
    pub fn add_gas_usage_in_shard(&mut self, shard_uid: ShardUId, shard_usage: GasUsageInShard) {
        let current_shard_usage =
            self.shards.entry(shard_uid).or_insert_with(|| GasUsageInShard::new());
        current_shard_usage.merge(&shard_usage);
    }
```
I don't really like this version. I feel like the current one is more readable and it avoids copying all elements from one map to the other during `merge`.